### ISSUE TITLE
Add site-level announcements to OJS

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#9253]
**Describe the problem you would like to solve**
Multi-journal instances of OJS want to be able to add announcements at the site-level.

**Describe the solution you'd like**
Add announcements to the admin settings area, so that administrators can add, edit and delete announcements that appear on the site-wide homepage.

The ability to view the latest announcements will be added to the homepage in the default theme. Each site-level announcement will have its own page.

**Who is asking for this feature?**
This is part of sponsored development work for a PKP|PS client, UNIANDES, who will use this feature in their custom theme.

## Implementation

This feature is hidden behind a config flag for 3.4.x and requires a migration to be run manually. Follow these steps to test:

1. Check out the PRs.
2. Add the following to the bottom of your `config.inc.php` file.
```
[features]
site_announcements = On
```
3. Run the following command to run the migration.
```
php lib/pkp/tools/migration.php \\PKP\\migration\\upgrade\\v3_5_0\\I9253_SiteAnnouncements up
```
4. Login as an admin.
5. Go to **Administration > Hosted Journals** and make sure there are two or more journals.
6. Go to **Administration > Site Settings > Announcements**.
7. Click the checkbox to enable announcements and enter a positive number in the **Display on Homepage** field.
8. Click the **Save** button.

Notes:
- When an announcement is sent at the journal level, an email is sent to all users who have subscribed to the journal's announcements. No emails are sent for announcements at the site level because users can not subscribe or unsubscribe to these announcements.
- This PR changes how announcements are ordered in the backend and frontend. Now they are shown with the latest first, by `date_posted` in descending order.
- This PR fixes a small visual bug in the list of announcements, due to a CSS override, that caused extra padding on the page that lists announcements.
- This only adjusts the default theme. Other themes would need to be modified separately to support this feature.
- This PR also adds support for a featured image with each announcement.

PRs:

**stable-3_4_0**:
https://github.com/pkp/pkp-lib/pull/9255
https://github.com/pkp/ui-library/pull/283
https://github.com/pkp/ojs/pull/4023

**main**:
https://github.com/pkp/pkp-lib/pull/9260
https://github.com/pkp/ui-library/pull/284
https://github.com/pkp/ojs/pull/4028

**Comment by NateWr** [comment]
The following PRs add support for announcements at the site level to the `stable-3_4_0` branch. Once approved, I will port to the other apps and to the `main` branch. @asmecher can someone review this?

https://github.com/pkp/pkp-lib/pull/9255
https://github.com/pkp/ojs/pull/4023

This feature is hidden behind a config flag for 3.4.x and requires a migration to be run manually. Follow these steps to test:

1. Check out the PRs.
2. Add the following to the bottom of your `config.inc.php` file.
```
[features]
site_announcements = On
```
3. Run the following command to run the migration.
```
php lib/pkp/tools/migration.php \\PKP\\migration\\upgrade\\v3_5_0\\I9253_SiteAnnouncements up
```
4. Login as an admin.
5. Go to **Administration > Hosted Journals** and make sure there are two or more journals.
6. Go to **Administration > Site Settings > Announcements**.
7. Click the checkbox to enable announcements and enter a positive number in the **Display on Homepage" field.
8. Click the **Save** button.

Notes:
- When an announcement is sent at the journal level, an email is sent to all users who have subscribed to the journal's announcements. No emails are sent for announcements at the site level because users can not subscribe or unsubscribe to these announcements.
- This PR changes how announcements are ordered in the backend and frontend. Now they are shown with the latest first, by `date_posted` in descending order.
- This PR fixes a small visual bug in the list of announcements, due to a CSS override, that caused extra padding on the page that lists announcements.
- This only adjusts the default theme. Other themes would need to be modified separately to support this feature.

**Comment by asmecher** [comment]
Reviews added -- thanks, @NateWr!

**Comment by NateWr** [comment]
Thanks @asmecher! I've made the change from `CONTEXT_SITE_ID` to `null` for site-level announcements and addressed your other comments. I also added a missing PR for ui-library -- just a small adjustment to show/hide stuff when announcements are enabled/disabled at the site level. There's just one unresolved comment regarding the interface for Site/Context classes.

Can you review this again when you have a moment?

**Comment by asmecher** [comment]
Looks good, @NateWr -- would you like this merged now, or would you rather do the porting to `main` first?

**Comment by NateWr** [comment]
Thanks @asmecher! Here are PRs for all apps for `stable-3_4_0` and `main`.

**stable_3-4-0**
- https://github.com/pkp/pkp-lib/pull/9255
- https://github.com/pkp/ui-library/pull/283
- https://github.com/pkp/ojs/pull/4023
- https://github.com/pkp/omp/pull/1452
- https://github.com/pkp/ops/pull/564

**main**
- https://github.com/pkp/pkp-lib/pull/9260
- https://github.com/pkp/ui-library/pull/284
- https://github.com/pkp/ojs/pull/4028
- https://github.com/pkp/omp/pull/1453
- https://github.com/pkp/ops/pull/565

**Comment by NateWr** [comment]
I found a small bug related to the URL of announcements in the API response. The key fix is making the `Context` optional in the `PKP\core\maps\Schema::__construct()` so that the dependency injector will pass `null` when at the site-level.

Otherwise, the PRs above remain as you last saw them.

**Comment by asmecher** [comment]
@NateWr, it looked like none of the tests passed; have you looked into that?

**Comment by NateWr** [comment]
Woops! That was quite a catastrophic minor change. Let's see if this will do it. :)

**Comment by touhidurabir** [comment]
@NateWr `OMP` specific changes for pkp/pkp-lib#8025 has been merged that was causing test failing . Please try now .

**Comment by defstat** [comment]
Not sure, but it seems that there are some merge conflicts regarding your PRs I think.

```
From https://github.com/NateWr/pkp-lib
 * branch                i9253_site_announcements_main -> FETCH_HEAD
 * [new branch]          i9253_site_announcements_main -> NateWr/i9253_site_announcements_main
Rebasing (1/16)
Auto-merging classes/invitation/invitations/BaseInvitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/invitations/BaseInvitation.php
Auto-merging classes/invitation/invitations/enums/InvitationStatus.php
CONFLICT (add/add): Merge conflict in classes/invitation/invitations/enums/InvitationStatus.php
Auto-merging classes/invitation/models/Invitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/models/Invitation.php
Auto-merging classes/invitation/repositories/BaseRepository.php
CONFLICT (add/add): Merge conflict in classes/invitation/repositories/BaseRepository.php
Auto-merging classes/invitation/repositories/Invitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/repositories/Invitation.php
Auto-merging pages/invitation/PKPInvitationHandler.php
CONFLICT (add/add): Merge conflict in pages/invitation/PKPInvitationHandler.php
Auto-merging pages/invitation/index.php
CONFLICT (add/add): Merge conflict in pages/invitation/index.php
error: could not apply ccb2123d8... pkp/pkp-lib#9197 Initial Commit
[33mhint: Resolve all conflicts manually, mark them as resolved with[m
[33mhint: "git add/rm <conflicted_files>", then run "git rebase --continue".[m
[33mhint: You can instead skip this commit: run "git rebase --skip".[m
[33mhint: To abort and get back to the state before "git rebase", run "git rebase --abort".[m
Could not apply ccb2123d8... pkp/pkp-lib#9197 Initial Commit
```

**Comment by NateWr** [comment]
@defstat I rebased the OJS pr against `main`, but I'm still running into test failures due to a syntax error in `BaseInvitation`. The pkp-lib PR is showing [no merge conflicts](https://github.com/pkp/pkp-lib/pull/9260). When I look at the OJS PR, I can see that `BaseInvitation` has [not been modified](https://github.com/pkp/ojs/pull/4028/files#diff-c19afe164738139d345a940eb083ada822cb4e7e34a6d11bbdf91e1dcd84265f). I'm not sure what to do here...

**Comment by defstat** [comment]
@NateWr I am in the process of checking what could be the problem on that. Thanks

**Comment by defstat** [comment]
I fetched and pulled everything (ojs, lib-pkp) and I created a new PR with a comment change on lib-pkp and one for OJS only for the submodule update (https://github.com/pkp/ojs/pull/4037, https://github.com/pkp/pkp-lib/pull/9292). 

I think that the error is perhaps related to the following lines of the [build](https://app.travis-ci.com/github/pkp/ojs/jobs/609517334):

```
7 - Updating lib/pkp with code from defstat repository, testInvitation branch.
HEAD is now at cc3c4895d pkp/pkp-lib#9197 ScheduledJob Fix
remote: Enumerating objects: 11, done.[K
remote: Counting objects:   9% (1/11)[K
remote: Counting objects:  18% (2/11)[K
remote: Counting objects:  27% (3/11)[K
remote: Counting objects:  36% (4/11)[K
remote: Counting objects:  45% (5/11)[K
remote: Counting objects:  54% (6/11)[K
remote: Counting objects:  63% (7/11)[K
remote: Counting objects:  72% (8/11)[K
remote: Counting objects:  81% (9/11)[K
remote: Counting objects:  90% (10/11)[K
remote: Counting objects: 100% (11/11)[K
remote: Counting objects: 100% (11/11), done.[K
remote: Compressing objects:  50% (1/2)[K
remote: Compressing objects: 100% (2/2)[K
remote: Compressing objects: 100% (2/2), done.[K
remote: Total 6 (delta 4), reused 6 (delta 4), pack-reused 0[K
Unpacking objects:  16% (1/6)
Unpacking objects:  33% (2/6)
Unpacking objects:  50% (3/6)
Unpacking objects:  66% (4/6)
Unpacking objects:  83% (5/6)
Unpacking objects: 100% (6/6)
Unpacking objects: 100% (6/6), 535 bytes | 267.00 KiB/s, done.
From https://github.com/defstat/pkp-lib
 * branch                testInvitation -> FETCH_HEAD
 * [new branch]          testInvitation -> defstat/testInvitation
Rebasing (1/16)
Auto-merging classes/invitation/invitations/BaseInvitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/invitations/BaseInvitation.php
Auto-merging classes/invitation/invitations/enums/InvitationStatus.php
CONFLICT (add/add): Merge conflict in classes/invitation/invitations/enums/InvitationStatus.php
Auto-merging classes/invitation/models/Invitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/models/Invitation.php
Auto-merging classes/invitation/repositories/BaseRepository.php
CONFLICT (add/add): Merge conflict in classes/invitation/repositories/BaseRepository.php
Auto-merging classes/invitation/repositories/Invitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/repositories/Invitation.php
Auto-merging pages/invitation/PKPInvitationHandler.php
CONFLICT (add/add): Merge conflict in pages/invitation/PKPInvitationHandler.php
Auto-merging pages/invitation/index.php
CONFLICT (add/add): Merge conflict in pages/invitation/index.php
error: could not apply ccb2123d8... pkp/pkp-lib#9197 Initial Commit
[33mhint: Resolve all conflicts manually, mark them as resolved with[m
[33mhint: "git add/rm <conflicted_files>", then run "git rebase --continue".[m
[33mhint: You can instead skip this commit: run "git rebase --skip".[m
[33mhint: To abort and get back to the state before "git rebase", run "git rebase --abort".[m
Could not apply ccb2123d8... pkp/pkp-lib#9197 Initial Commit
```
The same lines are found in your [build](https://app.travis-ci.com/github/pkp/ojs/jobs/609399841) also.

```
7 - Updating lib/pkp with code from NateWr repository, i9253_site_announcements_main branch.
HEAD is now at cc3c4895d pkp/pkp-lib#9197 ScheduledJob Fix
remote: ....... [DELETED LINES OF OUTPUT]
From https://github.com/NateWr/pkp-lib
 * branch                i9253_site_announcements_main -> FETCH_HEAD
 * [new branch]          i9253_site_announcements_main -> NateWr/i9253_site_announcements_main
Rebasing (1/16)
Auto-merging classes/invitation/invitations/BaseInvitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/invitations/BaseInvitation.php
Auto-merging classes/invitation/invitations/enums/InvitationStatus.php
CONFLICT (add/add): Merge conflict in classes/invitation/invitations/enums/InvitationStatus.php
Auto-merging classes/invitation/models/Invitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/models/Invitation.php
Auto-merging classes/invitation/repositories/BaseRepository.php
CONFLICT (add/add): Merge conflict in classes/invitation/repositories/BaseRepository.php
Auto-merging classes/invitation/repositories/Invitation.php
CONFLICT (add/add): Merge conflict in classes/invitation/repositories/Invitation.php
Auto-merging pages/invitation/PKPInvitationHandler.php
CONFLICT (add/add): Merge conflict in pages/invitation/PKPInvitationHandler.php
Auto-merging pages/invitation/index.php
CONFLICT (add/add): Merge conflict in pages/invitation/index.php
error: could not apply ccb2123d8... pkp/pkp-lib#9197 Initial Commit
[33mhint: Resolve all conflicts manually, mark them as resolved with[m
[33mhint: "git add/rm <conflicted_files>", then run "git rebase --continue".[m
[33mhint: You can instead skip this commit: run "git rebase --skip".[m
[33mhint: To abort and get back to the state before "git rebase", run "git rebase --abort".[m
Could not apply ccb2123d8... pkp/pkp-lib#9197 Initial Commit
```

It seems that, even though the main branch code has long passed `HEAD is now at cc3c4895d pkp/pkp-lib#9197 ScheduledJob Fix`, for some reason the build tries to rebase from that point on, resulting a merge conflict, and possibly the error you get.

Not sure what is causing that - Could it be a "faulty" module reference in the main branch? @asmecher do you have any ideas on this?

**Comment by NateWr** [comment]
@asmecher I'm now running into this same test failure on a separate pull request related to a different issue. For #9262 (features/highlights), the [OPS pull request](https://github.com/pkp/ops/pull/567/) is [failing](https://app.travis-ci.com/github/pkp/ops/jobs/610005275#L421) with the same error as @defstat indicated above.

For the same issue, the pull requests for [OJS](https://github.com/pkp/ojs/pull/4038) and [OMP](https://github.com/pkp/omp/pull/1458) are passing. All three pull requests point to the same commit in lib/pkp, but OPS's [commit hash for lib/pkp](https://github.com/pkp/pkp-lib/compare/cc3c4895db3c27cbe5c9f4a40889710a2988c392...cc2688ad4523da32d0674b59db137c1db18f4a37) is further behind [OMP's](https://github.com/pkp/pkp-lib/compare/733204eae82633fc84879bac6b4a2fd2cc13c8fa...cc2688ad4523da32d0674b59db137c1db18f4a37).

 I'm not able to reproduce the test failures locally. Any ideas? All other tests are passing on this issue and #9262, so this is now a blocker for all of the core contributions related to the PS work.

**Comment by asmecher** [comment]
@NateWr, it sounds like the issues with OJS and OMP that you were having above were resolved? I did speak with Jarda about something similar and push up submodule updates during the last week, but probably didn't do it for OPS. I've just done a submodule update on `pkp-lib` and `ui-library` for OPS `main`, and hopefully that'll resolve it. Please let me know! (Tests just started.)

**Comment by NateWr** [comment]
Ok, thanks everyone. All of the tests have passed now. The stable PRs went stale so I've rebased them, but assuming the tests will pass. I've moved the testing instructions to the first comment in this issue. Final PRs for review:

**stable-3_4_0**:
https://github.com/pkp/pkp-lib/pull/9255
https://github.com/pkp/ui-library/pull/283
https://github.com/pkp/ojs/pull/4023

**main**:
https://github.com/pkp/pkp-lib/pull/9260
https://github.com/pkp/ui-library/pull/284
https://github.com/pkp/ojs/pull/4028

**Comment by asmecher** [comment]
Thanks, @NateWr! I've added [a few small comments](https://github.com/pkp/pkp-lib/pull/9255#pullrequestreview-1649828036) -- the type checking definitely does need a quick look. Otherwise I think there's just a bit of minor porting to the other apps, and it's ready to merge!

**Comment by NateWr** [comment]
> just a bit of minor porting to the other apps

Makes total sense. The original brief was just for OJS, though. How do you feel about porting to other apps for `main` but not `stable-3_4_0`?

**Comment by asmecher** [comment]
Sorry for the wait on this, @NateWr. Yes, that's fine with me.

**Comment by NateWr** [comment]
@asmecher tests are passing for `stable-3_4_0` and `main`. I'd like to port to OMP and OPS, but work on those apps is blocked by this issue: https://github.com/pkp/pkp-lib/issues/9481

**Comment by asmecher** [comment]
Thanks, I've pinged Jarda on that one -- it shouldn't take long to resolve, and once that's done, can you tag me again here?

**Comment by NateWr** [comment]
@asmecher thanks, Jarda resolved that quickly yesterday. I've now got `main` ports for all apps and everything is passing. All PRs are ready to merge here: https://github.com/pkp/pkp-lib/issues/9253#issuecomment-1733476667

**Comment by asmecher** [comment]
All merged or cherry-picked, @NateWr! Can this be closed?
